File: mobile/native/darwin/Packages/EnteNetwork/Sources/EnteNetwork/Core/APIClient.swift
import Foundation
import EnteCore
import Logging

// MARK: - API Client

public class APIClient {
    private let httpClient: HTTPClient
    private let endpointResolver: EndpointResolver
    private let headersManager: RequestHeadersManager
    private let authTokenProvider: AuthTokenProvider?
    private let app: EnteApp
    private let logger = Logger(label: "APIClient")
    
    public init(
        configuration: NetworkConfiguration,
        app: EnteApp,
        authTokenProvider: AuthTokenProvider? = nil,
        httpClient: HTTPClient = HTTPClient()
    ) {
        self.httpClient = httpClient
        self.endpointResolver = EndpointResolver(configuration: configuration)
        self.headersManager = RequestHeadersManager()
        self.authTokenProvider = authTokenProvider
        self.app = app
    }
    
    public func request<T: Codable>(_ endpoint: APIEndpoint, responseType: T.Type) async throws -> T {
        let url = endpointResolver.resolveURL(for: endpoint)
        let headers = await buildHeaders(for: endpoint)
        
        logger.debug("Making request to \(url)")
        
        return try await httpClient.request(
            url: url,
            method: endpoint.method,
            parameters: endpoint.parameters,
            headers: headers,
            responseType: responseType
        )
    }
    
    public func request(_ endpoint: APIEndpoint) async throws -> EmptyResponse {
        return try await request(endpoint, responseType: EmptyResponse.self)
    }
    
    public func upload(_ endpoint: APIEndpoint, data: Data) async throws -> Data {
        let url = endpointResolver.resolveURL(for: endpoint)
        let headers = await buildHeaders(for: endpoint)
        
        return try await httpClient.upload(url: url, data: data, headers: headers)
    }
    
    public func download(_ endpoint: APIEndpoint) async throws -> Data {
        let url = endpointResolver.resolveURL(for: endpoint)
        let headers = await buildHeaders(for: endpoint)
        
        return try await httpClient.download(url: url, headers: headers)
    }
    
    private func buildHeaders(for endpoint: APIEndpoint) async -> [String: String] {
        // Start with endpoint-specific headers
        var headers = endpoint.headers ?? [:]
        
        // Add common headers (mirrors mobile network.dart pattern)
        let commonHeaders = await headersManager.buildHeaders(
            app: app,
            authTokenProvider: authTokenProvider
        )
        
        // Merge headers (endpoint headers take precedence)
        for (key, value) in commonHeaders {
            if headers[key] == nil {
                headers[key] = value
            }
        }
        
        return headers
    }
}

// MARK: - Simple Auth Token Provider

public class SimpleAuthTokenProvider: AuthTokenProvider {
    private var token: String?
    
    public init(token: String? = nil) {
        self.token = token
    }
    
    public func getAuthToken() async throws -> String? {
        return token
    }
    
    public func setAuthToken(_ token: String?) {
        self.token = token
    }
}

File: mobile/native/android/apps/ensu/data/src/main/java/io/ente/ensu/data/network/ApiClient.kt
package io.ente.ensu.data.network

import android.content.Context
import kotlinx.serialization.KSerializer

class ApiClient(
    context: Context,
    configuration: NetworkConfiguration,
    private val authTokenProvider: AuthTokenProvider? = null,
    private val httpClient: HttpClient = HttpClient()
) {
    private val endpointResolver = EndpointResolver(configuration)
    private val headersManager = RequestHeadersManager(context)

    suspend fun <T> request(endpoint: ApiEndpoint, serializer: KSerializer<T>): T {
        val url = endpointResolver.resolve(endpoint)
        val headers = headersManager.buildHeaders(authTokenProvider).toMutableMap()
        endpoint.headers?.forEach { (key, value) -> headers.putIfAbsent(key, value) }

        return httpClient.request(
            url = url,
            method = endpoint.method,
            parameters = endpoint.parameters,
            headers = headers,
            responseSerializer = serializer
        )
    }

    suspend fun request(endpoint: ApiEndpoint) {
        request(endpoint, EmptyResponse.serializer())
    }
}
